<a href="https://colab.research.google.com/github/HwangHanJae/Book-Recommendation/blob/main/EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# EDA

데이터를 확인하고 전처리 과정을 진행하겠습니다.

# 데이터 로드

In [3]:
import pandas as pd
import numpy as np

user_path = '/content/drive/MyDrive/Colab Notebooks/Recommendation/Book Recommendation/dataset/Users.csv'
book_path = '/content/drive/MyDrive/Colab Notebooks/Recommendation/Book Recommendation/dataset/Books.csv'
ratings_path = '/content/drive/MyDrive/Colab Notebooks/Recommendation/Book Recommendation/dataset/Ratings.csv'

users = pd.read_csv(user_path)
books = pd.read_csv(book_path)
ratings = pd.read_csv(ratings_path)

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


## 데이터 확인

### Users

In [68]:
users.head()

User-ID                            Location   Age   Location_split
0        1                  nyc, new york, usa   NaN              usa
1        2           stockton, california, usa  18.0              usa
2        3     moscow, yukon territory, russia   NaN           russia
3        4           porto, v.n.gaia, portugal  17.0         portugal
4        5  farnborough, hants, united kingdom   NaN   united kingdom

In [92]:
users.isnull().sum()

User-ID                0
Location               0
Age               110762
Location_split         0
dtype: int64

In [69]:
users['User-ID'].nunique()

278858

Users 데이터의 `User-ID`(사용자ID)는 결측치(NULL)는 없고 총 278858개가 있는 것을 확인할 수 있습니다.

### Books

In [71]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [93]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [73]:
books['ISBN'].nunique()

271360

Books 데이터의 `ISBN`(국제표준도서번호)는 결측치(NULL)는 없고 271360개가 있는 것을 확인할 수 있습니다.

### Ratings

In [85]:
ratings.shape

(1149780, 3)

In [78]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [66]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

ratings 데이터에는 결측치가 없는 것으로 확인됩니다.

In [79]:
ratings['User-ID'].nunique()

105283

In [80]:
ratings['ISBN'].nunique()

340556

ratings 데이터에서 `User-ID`의 `nunique`와 Users 데이터의 `User-ID`의 `nunique`가 다른것으로 확인됩니다.

또한 ratings 데이터의 `ISBN`과  books의 `ISBN`이 다른것으로 확인됩니다.

실제로 평점을 부여한 사용자와 평점이 부여된 아이템(책)만 가지고 추천을 이용하는 것이 좋겠습니다.

In [96]:
user_ratings_merge_df = pd.merge(ratings, users, how='inner', left_on = 'User-ID', right_on = 'User-ID')

user_ratings_merge_df.shape

(1149780, 6)

In [97]:
user_ratings_merge_df.head()

User-ID        ISBN  Book-Rating                       Location   Age  \
0   276725  034545104X            0              tyler, texas, usa   NaN   
1   276726  0155061224            5       seattle, washington, usa   NaN   
2   276727  0446520802            0  h, new south wales, australia  16.0   
3   276729  052165615X            3           rijeka, n/a, croatia  16.0   
4   276729  0521795028            6           rijeka, n/a, croatia  16.0   

  Location_split  
0            usa  
1            usa  
2      australia  
3        croatia  
4        croatia

위의 결과를 확인해보면 `User-ID`컬럼을 기준으로 데이터 프레임을 병합하였습니다.

In [99]:
pd.merge(user_ratings_merge_df, books, how='inner', left_on = 'ISBN', right_on = 'ISBN').isnull().sum()

User-ID                     0
ISBN                        0
Book-Rating                 0
Location                    0
Age                    277835
Location_split              0
Book-Title                  0
Book-Author                 1
Year-Of-Publication         0
Publisher                   2
Image-URL-S                 0
Image-URL-M                 0
Image-URL-L                 4
dtype: int64

In [90]:
user_ratings_merge_df.merge(books, left_on = 'ISBN', right_on = 'ISBN')

User-ID        ISBN  Book-Rating                           Location  \
0         276725  034545104X            0                  tyler, texas, usa   
1           2313  034545104X            5              cincinnati, ohio, usa   
2           6543  034545104X            0           strafford, missouri, usa   
3           8680  034545104X            5  st. charles county, missouri, usa   
4          10314  034545104X            9             beaverton, oregon, usa   
...          ...         ...          ...                                ...   
1031131   276688  0517145553            0          fort lee, new jersey, usa   
1031132   276688  1575660792            7          fort lee, new jersey, usa   
1031133   276690  0590907301            0                 wakeman, ohio, usa   
1031134   276704  0679752714            0             cedar park, texas, usa   
1031135   276704  0806917695            5             cedar park, texas, usa   

          Age Location_split  \
0         NaN            usa   
1        23.0            usa   
2        34.0            usa   
3         2.0            usa   
4         NaN            usa   
...       ...            ...   
1031131   NaN            usa   
1031132   NaN            usa   
1031133  43.0            usa   
1031134   NaN            usa   
1031135   NaN            usa   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104

# Model Class